docs/latex/Inhalt/03_Inhalt/07_refactoring.tex

```diff
@@ -0,0 +1,17 @@
+
+Aus diesen Gründen wurde beim Refactoring die Klasse \code{PersistenceManager} komplett entfernt und der Singleton statt dessen \enquote{implizit} umgesetzt, in dem in der Main-Klasse des Programms eine Instanz der \code{EntityManagerFactory} zentral erzeugt wird, welche dann in die Repositories injected wird. Diese Variante implementiert das Entwurfsmuster Singleton zwar nicht explizit, stellt jedoch tatsächlich sicher, dass nur eine Instanz der \code{EntityManagerFactory} existiert und ist außerdem threadsicher. Dieses Refactoring kann im Commit \href{https://github.com/anditru/quickie/commit/ab3ffd3a597ec6aee30ace0f75c9ace79617804e}{ab3ffd3a597ec6aee30ace0f75c9ace79617804e} eingesehen werden.
+
+\section{Doppelt vorhandener Code}
+Die Klasse \code{JPAProfileRepository} implementiert unter anderem die Methoden \code{persistentAdd}, \code{persistentUpdate} und \code{persistentRemove} zur Erstellung, Änderung und Löschung von Profiles. Da diese Methoden von der abstrakten Klasse \code{PersistentProfileRepository} geerbt werden, erwarten diese als Argument ein \code{PersistentProfile}. Die konkrete Implementierung \code{JPAProfileRepository} kann jedoch nur Profiles vom Typ \code{JPAProfile}, einer Unterklasse des \code{PersistentRecipe}, verarbeiten. Darum muss zu Beginn der Methoden geprüft werden, ob das übergebene Profile den passenden Typ besitzt, sonst wird eine Exception geworfen.
+
+Der hierfür zuständige Code ist in allen drei Methoden identisch, wurde jedoch drei explizit in jeder Methoden ausgeschrieben. Hierbei handelt es sich eindeutig um Duplicated Code, einen sehr kritischen Code Smell, der dazu führen würde, dass bei einer Änderung in den Anforderungen an das \code{JPAProfileRepository} die Änderungen ggf. an drei Stellen durchgeführt werden müssten. Aus diesem Grund wurde hier das Refactoring \enquote{Extract Method} angewendet und somit die Überprüfung des Typs in die Methode \code{assertInstanceOfJPARecipe} ausgelagert. Um die Struktur der Repositories im JPA-Plugin konsistent zu halten, wurde die entsprechende Typüberprüfung im \code{JPARecipeRepository} ebenfalls in eine separate Methode ausgelagert, obwohl sie hier nur einmal auftritt. Dieses Refactoring kann im Commit \href{https://github.com/anditru/quickie/commit/00a028943c4e6a02b5ad6f14f0268f60c9528e8f}{00a028943c4e6a02b5ad6f14f0268f60c9528e8f} eingesehen werden.
```

Comment:
" Der hierfür zuständige Code ist in allen drei Methoden identisch, wurde jedoch drei explizit in jeder Methoden ausgeschrieben" => drei?

Reply:
Ja, es waren drei Methoden, bei denen wir das überprüfen.

Reply:
"wurde jedoch drei explizit"

Reply:
Ups